File: runtime/core/agent.go
//core负责运行流程，domain定义数据
package core

import (
	"agent-runtime/domain"
	"context"
)

//防止修改原agent
type AgentSnapshot struct {
	ID           domain.ID            `json:"id"`
	Name         string               `json:"name"`
	Definition   domain.DefinitionRef `json:"definition"`
	Status       domain.AgentStatus   `json:"status"`
	State        map[string]any       `json:"state"`
	StateVersion uint64               `json:"state_version"`
	BindingError string               `json:"binding_error,omitempty"`
}

//本轮执行的输入
type ExecutionContext struct {
	Agent       AgentSnapshot `json:"agent"`
	Event       domain.Event  `json:"event"`
	ExecutionID domain.ID     `json:"execution_id"`
	AttemptID   domain.ID     `json:"attempt_id"`
}

//本轮输出，状态提交后才执行action
type ExecutionResult = domain.ExecutionResult

//agent执行接口
type AgentRunner interface {
	Run(context ExecutionContext) (ExecutionResult, error)
}

//支持取消的runner，原有接口仍可使用
type ContextAgentRunner interface {
	AgentRunner
	RunContext(context.Context, ExecutionContext) (ExecutionResult, error)
}

//适配层区分业务错误、运行错误和中断
type RunnerFailure interface {
	error
	FailureKind() domain.ErrorKind
}

File: python/src/agent_runtime/agent.py
"""agent输入快照与状态增量，持久化由go runtime负责"""

from dataclasses import dataclass, field
from decimal import Decimal
from typing import Protocol

type JSONValue = (
    None | bool | int | float | Decimal | str | list[JSONValue] | dict[str, JSONValue]
)
type JSONObject = dict[str, JSONValue]


class BusinessError(Exception):
    """本轮业务拒绝，不提交状态或action"""


@dataclass(frozen=True)
class DefinitionRef:
    id: str
    version: str


@dataclass(frozen=True)
class AgentSnapshot:
    id: str
    name: str
    definition: DefinitionRef
    status: str
    state: JSONObject | None
    state_version: int
    binding_error: str | None = None


@dataclass(frozen=True)
class Event:
    id: str
    type: str
    payload: JSONObject | None
    created_at: str


@dataclass(frozen=True)
class ExecutionContext:
    agent: AgentSnapshot
    event: Event
    execution_id: str
    attempt_id: str


@dataclass(frozen=True)
class Action:
    id: str
    type: str
    payload: JSONObject
    execution_id: str | None = None


@dataclass(frozen=True)
class ExecutionResult:
    state_update: JSONObject = field(default_factory=dict)
    actions: list[Action] = field(default_factory=list)


class Runner(Protocol):
    def run(self, context: ExecutionContext) -> ExecutionResult: ...
